pkg/oc/admin/diagnostics/diagnostics.go

```
@@ -274,14 +331,19 @@ func (o DiagnosticsOptions) RunDiagnostics() (bool, error, int, int) {
 				errors = append(errors, fmt.Errorf("While building the diagnostics, a panic was encountered.\nThis is a bug in diagnostics. Error and stack trace follow: \n%v\n%s", r, stack))
 			}
 		}()
-		detected, detectWarnings, detectErrors := o.detectClientConfig() // may log and return problems
+
+		// build client/cluster diags if there is a client config for them to use
+		expected, detected, detectWarnings, detectErrors := o.detectClientConfig() // may log and return problems
 		for _, warn := range detectWarnings {
 			warnings = append(warnings, warn)
 		}
 		for _, err := range detectErrors {
 			errors = append(errors, err)
 		}
-		if !detected { // there just plain isn't any client config file available
+		if !expected {
+			// no diagnostic required a client config, nothing to do
+		} else if !detected {
+			// there just plain isn't any client config file available
```

could this constitute a diagnostics failure (rather than just a log warning) if a clientConfig was expected but was not detected?

---

The general design of diagnostics has been to run all applicable diagnostics and skip the rest. So if you're missing a master config file, skip diagnostics that require that config file. Or if you have a client config file but aren't cluster-admin, skip the cluster diagnostics. It's not an error per se, just narrowing the scope of operation.

On the other hand, it would probably make sense to indicate failure if *all* diagnostics are skipped. With this PR it will be more common that only one diagnostic is requested, and if that's skipped and diagnostics happily declares success, that would be surprising.

---

> On the other hand, it would probably make sense to indicate failure if all diagnostics are skipped. 
> With this PR it will be more common that only one diagnostic is requested, and if that's skipped and diagnostics happily declares success, that would be surprising.

I agree, it would be reasonable to fail if all are skipped

---

I'm putting something in `func (o DiagnosticsOptions) Run()` to check for that.